asteroid/data/avspeech_dataset.py

```diff
@@ -0,0 +1,238 @@
+                                           stride=160))
+
+    @staticmethod
+    def encode(x: np.ndarray, p=0.3, stft_encoder=None):
+        if stft_encoder is None:
+            stft_encoder = Encoder(STFTFB(n_filters=512, kernel_size=400,
+                                          stride=160))
+
+        x = torch.from_numpy(x).float()
+
+        # time domain to time-frequency representation
+        tf_rep = stft_encoder(x).squeeze(0) + EPS
+        # power law on complex numbers
+        tf_rep = (torch.abs(tf_rep) ** p) * torch.sign(tf_rep)
+        # (514, 298) -> (257, 298, 2)
+        tf_rep = transforms.to_torchaudio(tf_rep)
+        return tf_rep
+
+    @staticmethod
+    def decode(tf_rep: np.ndarray, p=0.3, stft_decoder=None, final_len=48000):
+        if stft_decoder is None:
+            stft_decoder = Decoder(STFTFB(n_filters=512, kernel_size=400,
+                                          stride=160))
+
+        tf_rep = torch.from_numpy(tf_rep).float()
+
+        # (257, 298, 2) -> (514, 298)
+        tf_rep = transforms.from_torchaudio(tf_rep) + EPS
+        # power law on complex numbers
+        tf_rep = (torch.abs(tf_rep) ** (1/p)) * torch.sign(tf_rep)
+        # time domain to time-frequency representation
+        x = stft_decoder(tf_rep)
+
+        length = len(x)
+        if length != final_len:
+            x = F.pad(x, [0, final_len-length])
+        return x
+
+    def __len__(self):
+        return len(self.input_df)
+
+    def __getitem__(self, idx):
+        row = self.input_df.iloc[idx, :]
+        all_signals = []
+
+        for i in range(self.input_audio_size):
+            #get audio, video path from combination dataframe
+            video_path = row[i]
+            audio_path = row[i+self.input_audio_size]
+
+            #video length is 3-10 seconds, hence, part index can take values 0-2
+            re_match = re.search(r'_part\d', audio_path)
+            video_length_idx = 0
+            if re_match:
+                video_length_idx = int(re_match.group(0)[-1])
+
+            signal = Signal(video_path, audio_path, self.embed_dir,
+                            video_start_length=video_length_idx)
+            all_signals.append(signal)
+
+        #input audio signal is the last column.
+        mixed_signal, is_spec, spec_path = Signal.load_audio(row[-1])
+
+        if not is_spec:
+            mixed_signal = self.encode(mixed_signal, stft_encoder=self.stft_encoder)
+
+        audio_tensors = []
+        video_tensors = []
+
+        for i in range(self.input_audio_size):
+            #audio to spectrogram
+            if all_signals[i].spec_path.is_file():
+                spectrogram =  all_signals[i].get_spec()
+            else:
+                spectrogram = self.encode(all_signals[i].get_audio(),
+                                          stft_encoder=self.stft_encoder)
+            #convert to tensor
+            audio_tensors.append(spectrogram)
+
+            #check if the embedding is saved
```

Comment:
In the future, it would be nice to be able to do this on the fly and to cache the computed embeddings. 
But that's not for now I guess